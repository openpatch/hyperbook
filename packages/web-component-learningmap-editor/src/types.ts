import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface CompletionNeed {
  id: string;
  source?: string;
  target?: string;
}

export interface Completion {
  needs?: CompletionNeed[];
  optional?: CompletionNeed[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  background?: BackgroundConfig;
  edgeConfig?: EdgeConfig;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
