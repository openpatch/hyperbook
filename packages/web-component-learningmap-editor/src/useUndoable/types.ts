export type ActionType =
  | "undo"
  | "redo"
  | "update"
  | "reset"
  | "resetInitialState";

export type HistoryLimit = number | "infinium" | "infinity";

export type MutationBehavior =
  | "mergePastReversed"
  | "mergePast"
  | "destroyFuture"
  | "keepFuture";

export interface Action<T> {
  type: ActionType;
  payload?: T;
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
  ignoreAction?: boolean;
}

export interface State<T> {
  past: T[];
  present: T;
  future: T[];
}

export interface Options {
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
}

export type UseUndoable<T> = [
  T,
  (
    payload: T | ((oldValue: T) => T),
    behavior?: MutationBehavior,
    ignoreAction?: boolean,
  ) => void,
  {
    past: T[];
    future: T[];

    undo: () => void;
    canUndo: boolean;
    redo: () => void;
    canRedo: boolean;

    reset: (initialState?: T) => void;
    resetInitialState: (newInitialState: T) => void;
    static_setState: (
      payload: T,
      behavior?: MutationBehavior,
      ignoreAction?: boolean,
    ) => void;
  },
];
