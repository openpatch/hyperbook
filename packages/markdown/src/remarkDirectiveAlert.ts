// Register directive nodes in mdast:
/// <reference types="mdast-util-directive" />
//
import { HyperbookContext } from "@hyperbook/types";
import { Root } from "mdast";
import { visit } from "unist-util-visit";
import { VFile } from "vfile";
import {
  expectContainerDirective,
  isDirective,
  registerDirective,
} from "./remarkHelper";

export default (ctx: HyperbookContext) => () => {
  const name = "alert";
  return (tree: Root, file: VFile) => {
    visit(tree, function (node) {
      if (isDirective(node)) {
        if (node.name !== name) return;

        const data = node.data || (node.data = {});

        expectContainerDirective(node, file, name);
        registerDirective(file, name, [], ["style.css"], []);

        const color = node.attributes?.color;
        const content = node.attributes?.label;
        const type = Object.keys(node.attributes || {})
          .join(" ")
          .trim();
        const icon = content || type;

        delete node.attributes?.color;
        delete node.attributes?.label;
        data.hName = "div";
        data.hProperties = {
          class: `directive-alert ${type} ${icon ? "icon" : ""}`.trim(),
          style: `${color ? `--alert-color: ${color};` : ""} ${content ? `--alert-content: '${content}';` : ""}`,
        };
      }
    });
  };
};
