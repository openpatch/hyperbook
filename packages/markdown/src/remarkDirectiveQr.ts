// Register directive nodes in mdast:
/// <reference types="mdast-util-directive" />
//
import { HyperbookContext } from "@hyperbook/types";
import { Root } from "mdast";
import { visit } from "unist-util-visit";
import QRCode from "qrcode-svg";
import { VFile } from "vfile";
import {
  expectLeafDirective,
  isDirective,
  registerDirective,
} from "./remarkHelper";
import { ElementContent } from "hast";

const realSizes = {
  S: 64,
  M: 128,
  L: 256,
  XL: 512,
} as const;

export default (ctx: HyperbookContext) => () => {
  const name = "qr";
  return (tree: Root, file: VFile) => {
    visit(tree, function (node) {
      if (isDirective(node)) {
        if (node.name !== name) return;

        const data = node.data || (node.data = {});

        expectLeafDirective(node, file, name);
        registerDirective(file, name, [], ["style.css"]);

        const { value = "", size = "M", label } = node.attributes || {};

        const safeSize: "S" | "M" | "L" | "XL" = size as any;
        const qr = new QRCode({
          content: value as string,
          padding: 0,
          width: realSizes[safeSize],
          height: realSizes[safeSize],
          color: "#000000",
          background: "#ffffff",
          ecl: "M",
        }).svg();

        const labelElement: ElementContent[] = [];
        if (label) {
          labelElement.push({
            type: "element",
            tagName: "div",
            properties: {
              class: ["label", size].join(" "),
              style: `max-width: ${safeSize}px`,
            },
            children: [
              {
                type: "text",
                value: label,
              },
            ],
          });
        }
        data.hName = "div";
        data.hProperties = {
          class: "directive-qr",
        };
        data.hChildren = [
          {
            type: "element",
            tagName: "div",
            properties: {
              class: ["code", size].join(" "),
            },
            children: [
              {
                type: "raw",
                value: qr,
              },
            ],
          },
          ...labelElement,
        ];
      }
    });
  };
};
