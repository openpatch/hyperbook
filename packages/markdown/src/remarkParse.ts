// @ts-nocheck
import { Root } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import { Plugin } from "unified";
import micromarkSubSup from "./micromarkSubSup";
import mdastSubSup from "./mdastSubSup";

const remarkParse: Plugin<[], string, Root> = function () {
  const self = this;

  self.parser = function (doc) {
    const extensions = self.data("micromarkExtensions") || [];
    const mdastExtensions = self.data("fromMarkdownExtensions") || [];

    return fromMarkdown(doc, {
      ...self.data("settings"),
      extensions: [micromarkSubSup, ...extensions],
      mdastExtensions: [mdastSubSup, ...mdastExtensions],
    });
  };
};

export default remarkParse;
