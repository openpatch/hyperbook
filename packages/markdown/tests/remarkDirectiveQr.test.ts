import { HyperbookContext } from "@hyperbook/types/dist";
import { describe, expect, it } from "vitest";
import rehypeStringify from "rehype-stringify";
import remarkParse from "remark-parse";
import remarkToRehype from "remark-rehype";
import rehypeFormat from "rehype-format";
import { unified, PluggableList } from "unified";
import remarkDirective from "remark-directive";
import remarkDirectiveRehype from "remark-directive-rehype";
import { ctx } from "./mock";
import remarkDirectiveAlert from "../src/remarkDirectiveAlert";
import remarkDirectiveQr from "../src/remarkDirectiveQr";

export const toHtml = (md: string, ctx: HyperbookContext) => {
  const remarkPlugins: PluggableList = [
    remarkDirective,
    remarkDirectiveRehype,
    remarkDirectiveQr(ctx),
  ];

  return unified()
    .use(remarkParse)
    .use(remarkPlugins)
    .use(remarkToRehype)
    .use(rehypeFormat)
    .use(rehypeStringify, {
      allowDangerousCharacters: true,
      allowDangerousHtml: true,
    })
    .processSync(md);
};

describe("remarkDirectiveQr", () => {
  it("should transform", async () => {
    expect(
      toHtml(
        `
::qr{value="https://openpatch.org"}
`,
        ctx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should register directives", async () => {
    expect(
      toHtml(
        `
::qr{value="https://openpatch.org"}
`,
        ctx,
      ).data.directives?.["qr"],
    ).toBeDefined();
  });
});
