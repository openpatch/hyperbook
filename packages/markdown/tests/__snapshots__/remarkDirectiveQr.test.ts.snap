// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkDirectiveQr > should transform 1`] = `
"
<div class="directive-qr">
  <div class="code M"><?xml version="1.0" standalone="yes"?>
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" viewBox="0 0 128 128">
  <rect x="0" y="0" width="128" height="128" style="fill:#ffffff;shape-rendering:crispEdges;"/>
  <path x="0" y="0" style="fill:#000000;shape-rendering:crispEdges;" d="M0,0 V5.12 H5.12 V0 H0 Z M5.12,0 V5.12 H10.24 V0 H5.12 Z M10.24,0 V5.12 H15.36 V0 H10.24 Z M15.36,0 V5.12 H20.48 V0 H15.36 Z M20.48,0 V5.12 H25.60 V0 H20.48 Z M25.60,0 V5.12 H30.72 V0 H25.60 Z M30.72,0 V5.12 H35.84 V0 H30.72 Z M40.96,0 V5.12 H46.08 V0 H40.96 Z M51.20,0 V5.12 H56.32 V0 H51.20 Z M56.32,0 V5.12 H61.44 V0 H56.32 Z M66.56,0 V5.12 H71.68 V0 H66.56 Z M71.68,0 V5.12 H76.80 V0 H71.68 Z M76.80,0 V5.12 H81.92 V0 H76.80 Z M81.92,0 V5.12 H87.04 V0 H81.92 Z M92.16,0 V5.12 H97.28 V0 H92.16 Z M97.28,0 V5.12 H102.40 V0 H97.28 Z M102.40,0 V5.12 H107.52 V0 H102.40 Z M107.52,0 V5.12 H112.64 V0 H107.52 Z M112.64,0 V5.12 H117.76 V0 H112.64 Z M117.76,0 V5.12 H122.88 V0 H117.76 Z M122.88,0 V5.12 H128 V0 H122.88 Z M0,5.12 V10.24 H5.12 V5.12 H0 Z M30.72,5.12 V10.24 H35.84 V5.12 H30.72 Z M46.08,5.12 V10.24 H51.20 V5.12 H46.08 Z M51.20,5.12 V10.24 H56.32 V5.12 H51.20 Z M56.32,5.12 V10.24 H61.44 V5.12 H56.32 Z M71.68,5.12 V10.24 H76.80 V5.12 H71.68 Z M92.16,5.12 V10.24 H97.28 V5.12 H92.16 Z M122.88,5.12 V10.24 H128 V5.12 H122.88 Z M0,10.24 V15.36 H5.12 V10.24 H0 Z M10.24,10.24 V15.36 H15.36 V10.24 H10.24 Z M15.36,10.24 V15.36 H20.48 V10.24 H15.36 Z M20.48,10.24 V15.36 H25.60 V10.24 H20.48 Z M30.72,10.24 V15.36 H35.84 V10.24 H30.72 Z M46.08,10.24 V15.36 H51.20 V10.24 H46.08 Z M51.20,10.24 V15.36 H56.32 V10.24 H51.20 Z M81.92,10.24 V15.36 H87.04 V10.24 H81.92 Z M92.16,10.24 V15.36 H97.28 V10.24 H92.16 Z M102.40,10.24 V15.36 H107.52 V10.24 H102.40 Z M107.52,10.24 V15.36 H112.64 V10.24 H107.52 Z M112.64,10.24 V15.36 H117.76 V10.24 H112.64 Z M122.88,10.24 V15.36 H128 V10.24 H122.88 Z M0,15.36 V20.48 H5.12 V15.36 H0 Z M10.24,15.36 V20.48 H15.36 V15.36 H10.24 Z M15.36,15.36 V20.48 H20.48 V15.36 H15.36 Z M20.48,15.36 V20.48 H25.60 V15.36 H20.48 Z M30.72,15.36 V20.48 H35.84 V15.36 H30.72 Z M40.96,15.36 V20.48 H46.08 V15.36 H40.96 Z M46.08,15.36 V20.48 H51.20 V15.36 H46.08 Z M56.32,15.36 V20.48 H61.44 V15.36 H56.32 Z M61.44,15.36 V20.48 H66.56 V15.36 H61.44 Z M66.56,15.36 V20.48 H71.68 V15.36 H66.56 Z M71.68,15.36 V20.48 H76.80 V15.36 H71.68 Z M76.80,15.36 V20.48 H81.92 V15.36 H76.80 Z M92.16,15.36 V20.48 H97.28 V15.36 H92.16 Z M102.40,15.36 V20.48 H107.52 V15.36 H102.40 Z M107.52,15.36 V20.48 H112.64 V15.36 H107.52 Z M112.64,15.36 V20.48 H117.76 V15.36 H112.64 Z M122.88,15.36 V20.48 H128 V15.36 H122.88 Z M0,20.48 V25.60 H5.12 V20.48 H0 Z M10.24,20.48 V25.60 H15.36 V20.48 H10.24 Z M15.36,20.48 V25.60 H20.48 V20.48 H15.36 Z M20.48,20.48 V25.60 H25.60 V20.48 H20.48 Z M30.72,20.48 V25.60 H35.84 V20.48 H30.72 Z M40.96,20.48 V25.60 H46.08 V20.48 H40.96 Z M46.08,20.48 V25.60 H51.20 V20.48 H46.08 Z M51.20,20.48 V25.60 H56.32 V20.48 H51.20 Z M56.32,20.48 V25.60 H61.44 V20.48 H56.32 Z M61.44,20.48 V25.60 H66.56 V20.48 H61.44 Z M81.92,20.48 V25.60 H87.04 V20.48 H81.92 Z M92.16,20.48 V25.60 H97.28 V20.48 H92.16 Z M102.40,20.48 V25.60 H107.52 V20.48 H102.40 Z M107.52,20.48 V25.60 H112.64 V20.48 H107.52 Z M112.64,20.48 V25.60 H117.76 V20.48 H112.64 Z M122.88,20.48 V25.60 H128 V20.48 H122.88 Z M0,25.60 V30.72 H5.12 V25.60 H0 Z M30.72,25.60 V30.72 H35.84 V25.60 H30.72 Z M40.96,25.60 V30.72 H46.08 V25.60 H40.96 Z M51.20,25.60 V30.72 H56.32 V25.60 H51.20 Z M56.32,25.60 V30.72 H61.44 V25.60 H56.32 Z M61.44,25.60 V30.72 H66.56 V25.60 H61.44 Z M71.68,25.60 V30.72 H76.80 V25.60 H71.68 Z M76.80,25.60 V30.72 H81.92 V25.60 H76.80 Z M92.16,25.60 V30.72 H97.28 V25.60 H92.16 Z M122.88,25.60 V30.72 H128 V25.60 H122.88 Z M0,30.72 V35.84 H5.12 V30.72 H0 Z M5.12,30.72 V35.84 H10.24 V30.72 H5.12 Z M10.24,30.72 V35.84 H15.36 V30.72 H10.24 Z M15.36,30.72 V35.84 H20.48 V30.72 H15.36 Z M20.48,30.72 V35.84 H25.60 V30.72 H20.48 Z M25.60,30.72 V35.84 H30.72 V30.72 H25.60 Z M30.72,30.72 V35.84 H35.84 V30.72 H30.72 Z M40.96,30.72 V35.84 H46.08 V30.72 H40.96 Z M51.20,30.72 V35.84 H56.32 V30.72 H51.20 Z M61.44,30.72 V35.84 H66.56 V30.72 H61.44 Z M71.68,30.72 V35.84 H76.80 V30.72 H71.68 Z M81.92,30.72 V35.84 H87.04 V30.72 H81.92 Z M92.16,30.72 V35.84 H97.28 V30.72 H92.16 Z M97.28,30.72 V35.84 H102.40 V30.72 H97.28 Z M102.40,30.72 V35.84 H107.52 V30.72 H102.40 Z M107.52,30.72 V35.84 H112.64 V30.72 H107.52 Z M112.64,30.72 V35.84 H117.76 V30.72 H112.64 Z M117.76,30.72 V35.84 H122.88 V30.72 H117.76 Z M122.88,30.72 V35.84 H128 V30.72 H122.88 Z M40.96,35.84 V40.96 H46.08 V35.84 H40.96 Z M46.08,35.84 V40.96 H51.20 V35.84 H46.08 Z M51.20,35.84 V40.96 H56.32 V35.84 H51.20 Z M71.68,35.84 V40.96 H76.80 V35.84 H71.68 Z M0,40.96 V46.08 H5.12 V40.96 H0 Z M5.12,40.96 V46.08 H10.24 V40.96 H5.12 Z M10.24,40.96 V46.08 H15.36 V40.96 H10.24 Z M20.48,40.96 V46.08 H25.60 V40.96 H20.48 Z M30.72,40.96 V46.08 H35.84 V40.96 H30.72 Z M40.96,40.96 V46.08 H46.08 V40.96 H40.96 Z M46.08,40.96 V46.08 H51.20 V40.96 H46.08 Z M61.44,40.96 V46.08 H66.56 V40.96 H61.44 Z M66.56,40.96 V46.08 H71.68 V40.96 H66.56 Z M71.68,40.96 V46.08 H76.80 V40.96 H71.68 Z M81.92,40.96 V46.08 H87.04 V40.96 H81.92 Z M87.04,40.96 V46.08 H92.16 V40.96 H87.04 Z M92.16,40.96 V46.08 H97.28 V40.96 H92.16 Z M97.28,40.96 V46.08 H102.40 V40.96 H97.28 Z M102.40,40.96 V46.08 H107.52 V40.96 H102.40 Z M107.52,40.96 V46.08 H112.64 V40.96 H107.52 Z M122.88,40.96 V46.08 H128 V40.96 H122.88 Z M10.24,46.08 V51.20 H15.36 V46.08 H10.24 Z M15.36,46.08 V51.20 H20.48 V46.08 H15.36 Z M35.84,46.08 V51.20 H40.96 V46.08 H35.84 Z M40.96,46.08 V51.20 H46.08 V46.08 H40.96 Z M51.20,46.08 V51.20 H56.32 V46.08 H51.20 Z M61.44,46.08 V51.20 H66.56 V46.08 H61.44 Z M66.56,46.08 V51.20 H71.68 V46.08 H66.56 Z M76.80,46.08 V51.20 H81.92 V46.08 H76.80 Z M81.92,46.08 V51.20 H87.04 V46.08 H81.92 Z M87.04,46.08 V51.20 H92.16 V46.08 H87.04 Z M92.16,46.08 V51.20 H97.28 V46.08 H92.16 Z M97.28,46.08 V51.20 H102.40 V46.08 H97.28 Z M102.40,46.08 V51.20 H107.52 V46.08 H102.40 Z M107.52,46.08 V51.20 H112.64 V46.08 H107.52 Z M112.64,46.08 V51.20 H117.76 V46.08 H112.64 Z M10.24,51.20 V56.32 H15.36 V51.20 H10.24 Z M15.36,51.20 V56.32 H20.48 V51.20 H15.36 Z M30.72,51.20 V56.32 H35.84 V51.20 H30.72 Z M46.08,51.20 V56.32 H51.20 V51.20 H46.08 Z M61.44,51.20 V56.32 H66.56 V51.20 H61.44 Z M66.56,51.20 V56.32 H71.68 V51.20 H66.56 Z M87.04,51.20 V56.32 H92.16 V51.20 H87.04 Z M92.16,51.20 V56.32 H97.28 V51.20 H92.16 Z M107.52,51.20 V56.32 H112.64 V51.20 H107.52 Z M112.64,51.20 V56.32 H117.76 V51.20 H112.64 Z M117.76,51.20 V56.32 H122.88 V51.20 H117.76 Z M5.12,56.32 V61.44 H10.24 V56.32 H5.12 Z M15.36,56.32 V61.44 H20.48 V56.32 H15.36 Z M25.60,56.32 V61.44 H30.72 V56.32 H25.60 Z M35.84,56.32 V61.44 H40.96 V56.32 H35.84 Z M40.96,56.32 V61.44 H46.08 V56.32 H40.96 Z M61.44,56.32 V61.44 H66.56 V56.32 H61.44 Z M71.68,56.32 V61.44 H76.80 V56.32 H71.68 Z M76.80,56.32 V61.44 H81.92 V56.32 H76.80 Z M87.04,56.32 V61.44 H92.16 V56.32 H87.04 Z M92.16,56.32 V61.44 H97.28 V56.32 H92.16 Z M5.12,61.44 V66.56 H10.24 V61.44 H5.12 Z M15.36,61.44 V66.56 H20.48 V61.44 H15.36 Z M30.72,61.44 V66.56 H35.84 V61.44 H30.72 Z M51.20,61.44 V66.56 H56.32 V61.44 H51.20 Z M56.32,61.44 V66.56 H61.44 V61.44 H56.32 Z M61.44,61.44 V66.56 H66.56 V61.44 H61.44 Z M71.68,61.44 V66.56 H76.80 V61.44 H71.68 Z M87.04,61.44 V66.56 H92.16 V61.44 H87.04 Z M97.28,61.44 V66.56 H102.40 V61.44 H97.28 Z M102.40,61.44 V66.56 H107.52 V61.44 H102.40 Z M112.64,61.44 V66.56 H117.76 V61.44 H112.64 Z M10.24,66.56 V71.68 H15.36 V66.56 H10.24 Z M15.36,66.56 V71.68 H20.48 V66.56 H15.36 Z M20.48,66.56 V71.68 H25.60 V66.56 H20.48 Z M25.60,66.56 V71.68 H30.72 V66.56 H25.60 Z M35.84,66.56 V71.68 H40.96 V66.56 H35.84 Z M51.20,66.56 V71.68 H56.32 V66.56 H51.20 Z M66.56,66.56 V71.68 H71.68 V66.56 H66.56 Z M87.04,66.56 V71.68 H92.16 V66.56 H87.04 Z M102.40,66.56 V71.68 H107.52 V66.56 H102.40 Z M122.88,66.56 V71.68 H128 V66.56 H122.88 Z M0,71.68 V76.80 H5.12 V71.68 H0 Z M20.48,71.68 V76.80 H25.60 V71.68 H20.48 Z M25.60,71.68 V76.80 H30.72 V71.68 H25.60 Z M30.72,71.68 V76.80 H35.84 V71.68 H30.72 Z M40.96,71.68 V76.80 H46.08 V71.68 H40.96 Z M51.20,71.68 V76.80 H56.32 V71.68 H51.20 Z M66.56,71.68 V76.80 H71.68 V71.68 H66.56 Z M71.68,71.68 V76.80 H76.80 V71.68 H71.68 Z M81.92,71.68 V76.80 H87.04 V71.68 H81.92 Z M97.28,71.68 V76.80 H102.40 V71.68 H97.28 Z M102.40,71.68 V76.80 H107.52 V71.68 H102.40 Z M107.52,71.68 V76.80 H112.64 V71.68 H107.52 Z M112.64,71.68 V76.80 H117.76 V71.68 H112.64 Z M117.76,71.68 V76.80 H122.88 V71.68 H117.76 Z M122.88,71.68 V76.80 H128 V71.68 H122.88 Z M10.24,76.80 V81.92 H15.36 V76.80 H10.24 Z M15.36,76.80 V81.92 H20.48 V76.80 H15.36 Z M35.84,76.80 V81.92 H40.96 V76.80 H35.84 Z M40.96,76.80 V81.92 H46.08 V76.80 H40.96 Z M46.08,76.80 V81.92 H51.20 V76.80 H46.08 Z M51.20,76.80 V81.92 H56.32 V76.80 H51.20 Z M61.44,76.80 V81.92 H66.56 V76.80 H61.44 Z M66.56,76.80 V81.92 H71.68 V76.80 H66.56 Z M71.68,76.80 V81.92 H76.80 V76.80 H71.68 Z M81.92,76.80 V81.92 H87.04 V76.80 H81.92 Z M97.28,76.80 V81.92 H102.40 V76.80 H97.28 Z M102.40,76.80 V81.92 H107.52 V76.80 H102.40 Z M112.64,76.80 V81.92 H117.76 V76.80 H112.64 Z M117.76,76.80 V81.92 H122.88 V76.80 H117.76 Z M0,81.92 V87.04 H5.12 V81.92 H0 Z M10.24,81.92 V87.04 H15.36 V81.92 H10.24 Z M15.36,81.92 V87.04 H20.48 V81.92 H15.36 Z M20.48,81.92 V87.04 H25.60 V81.92 H20.48 Z M30.72,81.92 V87.04 H35.84 V81.92 H30.72 Z M40.96,81.92 V87.04 H46.08 V81.92 H40.96 Z M66.56,81.92 V87.04 H71.68 V81.92 H66.56 Z M76.80,81.92 V87.04 H81.92 V81.92 H76.80 Z M81.92,81.92 V87.04 H87.04 V81.92 H81.92 Z M87.04,81.92 V87.04 H92.16 V81.92 H87.04 Z M92.16,81.92 V87.04 H97.28 V81.92 H92.16 Z M97.28,81.92 V87.04 H102.40 V81.92 H97.28 Z M102.40,81.92 V87.04 H107.52 V81.92 H102.40 Z M107.52,81.92 V87.04 H112.64 V81.92 H107.52 Z M122.88,81.92 V87.04 H128 V81.92 H122.88 Z M40.96,87.04 V92.16 H46.08 V87.04 H40.96 Z M46.08,87.04 V92.16 H51.20 V87.04 H46.08 Z M51.20,87.04 V92.16 H56.32 V87.04 H51.20 Z M56.32,87.04 V92.16 H61.44 V87.04 H56.32 Z M71.68,87.04 V92.16 H76.80 V87.04 H71.68 Z M76.80,87.04 V92.16 H81.92 V87.04 H76.80 Z M81.92,87.04 V92.16 H87.04 V87.04 H81.92 Z M102.40,87.04 V92.16 H107.52 V87.04 H102.40 Z M107.52,87.04 V92.16 H112.64 V87.04 H107.52 Z M112.64,87.04 V92.16 H117.76 V87.04 H112.64 Z M122.88,87.04 V92.16 H128 V87.04 H122.88 Z M0,92.16 V97.28 H5.12 V92.16 H0 Z M5.12,92.16 V97.28 H10.24 V92.16 H5.12 Z M10.24,92.16 V97.28 H15.36 V92.16 H10.24 Z M15.36,92.16 V97.28 H20.48 V92.16 H15.36 Z M20.48,92.16 V97.28 H25.60 V92.16 H20.48 Z M25.60,92.16 V97.28 H30.72 V92.16 H25.60 Z M30.72,92.16 V97.28 H35.84 V92.16 H30.72 Z M51.20,92.16 V97.28 H56.32 V92.16 H51.20 Z M56.32,92.16 V97.28 H61.44 V92.16 H56.32 Z M61.44,92.16 V97.28 H66.56 V92.16 H61.44 Z M71.68,92.16 V97.28 H76.80 V92.16 H71.68 Z M76.80,92.16 V97.28 H81.92 V92.16 H76.80 Z M81.92,92.16 V97.28 H87.04 V92.16 H81.92 Z M92.16,92.16 V97.28 H97.28 V92.16 H92.16 Z M102.40,92.16 V97.28 H107.52 V92.16 H102.40 Z M107.52,92.16 V97.28 H112.64 V92.16 H107.52 Z M0,97.28 V102.40 H5.12 V97.28 H0 Z M30.72,97.28 V102.40 H35.84 V97.28 H30.72 Z M46.08,97.28 V102.40 H51.20 V97.28 H46.08 Z M51.20,97.28 V102.40 H56.32 V97.28 H51.20 Z M56.32,97.28 V102.40 H61.44 V97.28 H56.32 Z M61.44,97.28 V102.40 H66.56 V97.28 H61.44 Z M76.80,97.28 V102.40 H81.92 V97.28 H76.80 Z M81.92,97.28 V102.40 H87.04 V97.28 H81.92 Z M102.40,97.28 V102.40 H107.52 V97.28 H102.40 Z M117.76,97.28 V102.40 H122.88 V97.28 H117.76 Z M0,102.40 V107.52 H5.12 V102.40 H0 Z M10.24,102.40 V107.52 H15.36 V102.40 H10.24 Z M15.36,102.40 V107.52 H20.48 V102.40 H15.36 Z M20.48,102.40 V107.52 H25.60 V102.40 H20.48 Z M30.72,102.40 V107.52 H35.84 V102.40 H30.72 Z M40.96,102.40 V107.52 H46.08 V102.40 H40.96 Z M46.08,102.40 V107.52 H51.20 V102.40 H46.08 Z M56.32,102.40 V107.52 H61.44 V102.40 H56.32 Z M66.56,102.40 V107.52 H71.68 V102.40 H66.56 Z M71.68,102.40 V107.52 H76.80 V102.40 H71.68 Z M81.92,102.40 V107.52 H87.04 V102.40 H81.92 Z M87.04,102.40 V107.52 H92.16 V102.40 H87.04 Z M92.16,102.40 V107.52 H97.28 V102.40 H92.16 Z M97.28,102.40 V107.52 H102.40 V102.40 H97.28 Z M102.40,102.40 V107.52 H107.52 V102.40 H102.40 Z M112.64,102.40 V107.52 H117.76 V102.40 H112.64 Z M117.76,102.40 V107.52 H122.88 V102.40 H117.76 Z M0,107.52 V112.64 H5.12 V107.52 H0 Z M10.24,107.52 V112.64 H15.36 V107.52 H10.24 Z M15.36,107.52 V112.64 H20.48 V107.52 H15.36 Z M20.48,107.52 V112.64 H25.60 V107.52 H20.48 Z M30.72,107.52 V112.64 H35.84 V107.52 H30.72 Z M46.08,107.52 V112.64 H51.20 V107.52 H46.08 Z M51.20,107.52 V112.64 H56.32 V107.52 H51.20 Z M56.32,107.52 V112.64 H61.44 V107.52 H56.32 Z M71.68,107.52 V112.64 H76.80 V107.52 H71.68 Z M76.80,107.52 V112.64 H81.92 V107.52 H76.80 Z M112.64,107.52 V112.64 H117.76 V107.52 H112.64 Z M122.88,107.52 V112.64 H128 V107.52 H122.88 Z M0,112.64 V117.76 H5.12 V112.64 H0 Z M10.24,112.64 V117.76 H15.36 V112.64 H10.24 Z M15.36,112.64 V117.76 H20.48 V112.64 H15.36 Z M20.48,112.64 V117.76 H25.60 V112.64 H20.48 Z M30.72,112.64 V117.76 H35.84 V112.64 H30.72 Z M40.96,112.64 V117.76 H46.08 V112.64 H40.96 Z M46.08,112.64 V117.76 H51.20 V112.64 H46.08 Z M56.32,112.64 V117.76 H61.44 V112.64 H56.32 Z M71.68,112.64 V117.76 H76.80 V112.64 H71.68 Z M76.80,112.64 V117.76 H81.92 V112.64 H76.80 Z M81.92,112.64 V117.76 H87.04 V112.64 H81.92 Z M87.04,112.64 V117.76 H92.16 V112.64 H87.04 Z M112.64,112.64 V117.76 H117.76 V112.64 H112.64 Z M117.76,112.64 V117.76 H122.88 V112.64 H117.76 Z M0,117.76 V122.88 H5.12 V117.76 H0 Z M30.72,117.76 V122.88 H35.84 V117.76 H30.72 Z M40.96,117.76 V122.88 H46.08 V117.76 H40.96 Z M46.08,117.76 V122.88 H51.20 V117.76 H46.08 Z M56.32,117.76 V122.88 H61.44 V117.76 H56.32 Z M66.56,117.76 V122.88 H71.68 V117.76 H66.56 Z M71.68,117.76 V122.88 H76.80 V117.76 H71.68 Z M81.92,117.76 V122.88 H87.04 V117.76 H81.92 Z M87.04,117.76 V122.88 H92.16 V117.76 H87.04 Z M97.28,117.76 V122.88 H102.40 V117.76 H97.28 Z M102.40,117.76 V122.88 H107.52 V117.76 H102.40 Z M107.52,117.76 V122.88 H112.64 V117.76 H107.52 Z M112.64,117.76 V122.88 H117.76 V117.76 H112.64 Z M117.76,117.76 V122.88 H122.88 V117.76 H117.76 Z M0,122.88 V128 H5.12 V122.88 H0 Z M5.12,122.88 V128 H10.24 V122.88 H5.12 Z M10.24,122.88 V128 H15.36 V122.88 H10.24 Z M15.36,122.88 V128 H20.48 V122.88 H15.36 Z M20.48,122.88 V128 H25.60 V122.88 H20.48 Z M25.60,122.88 V128 H30.72 V122.88 H25.60 Z M30.72,122.88 V128 H35.84 V122.88 H30.72 Z M40.96,122.88 V128 H46.08 V122.88 H40.96 Z M51.20,122.88 V128 H56.32 V122.88 H51.20 Z M56.32,122.88 V128 H61.44 V122.88 H56.32 Z M61.44,122.88 V128 H66.56 V122.88 H61.44 Z M71.68,122.88 V128 H76.80 V122.88 H71.68 Z M76.80,122.88 V128 H81.92 V122.88 H76.80 Z M92.16,122.88 V128 H97.28 V122.88 H92.16 Z M97.28,122.88 V128 H102.40 V122.88 H97.28 Z M102.40,122.88 V128 H107.52 V122.88 H102.40 Z M107.52,122.88 V128 H112.64 V122.88 H107.52 Z M112.64,122.88 V128 H117.76 V122.88 H112.64 Z M117.76,122.88 V128 H122.88 V122.88 H117.76 Z M122.88,122.88 V128 H128 V122.88 H122.88 Z " /></svg></div>
</div>
"
`;
